contributing/code-style-and-quality-general.md

```diff
@@ -0,0 +1,667 @@
+
+While it is hard to exactly specify what constitutes a good design, there are some properties that can serve as a _litmus test_ for a good design. If these properties are given, the chances are good that the design is going into a good direction. If these properties cannot be achieved, there is a high probability that the design is flawed.
+
+
+### Immutability and Eager Initialization
+
+1. Try to use immutable types where possible, especially for APIs, messages, identifiers, properties, configuration, etc.
+2. A good general approach is to try and make as many fields of a class `final` as possible.
+3. Classes that are used as keys in maps should be strictly immutable and only have `final` fields (except maybe auxiliary fields, like lazy cached hash codes).
+4. Eagerly initialize classes. There should be no `init()` or `setup()` methods. Once the constructor completes, the object should be usable.
+
+
+### Nullability of the Mutable Parts
+
+For nullability, the Flink codebase aims to follow these conventions:
+
+* Fields, parameters, and return types are always non-null, unless indicated otherwise
+* All fields, parameters and method types that can be null should be annotated with `@javax.annotation.Nullable`. 
+That way, you get warnings from IntelliJ about all sections where you have to reason about potential null values.
+* For all mutable (non-final) fields that are not annotated, the assumption is that while the field value changes, there always is a value.
+    * This should be double check whether these can in fact not be null throughout the lifetime of the object.
+
+_Note: This means that `@Nonnull` annotations are usually not necessary, but can be used in certain cases to override a previous annotation, or to point non-nullability out in a context where one would expect a nullable value._
+
+`Optional` is a good solution as a return type for method that may or may not have a result, so nullable return types are good candidates to be replaced with `Optional`. 
+For fields and parameters, `Optional` is disputed in Java and most parts of the Flink code case don’t use optional for fields.
+
+
+### Avoid Code Duplication
+
+1. Whenever you are about to copy/paste some code, or reproduce a similar type of functionality in a different place, think about the ways how to refactor/reuse/abstract the changes to avoid the duplication.
+2. Common behavior between different specializations should be shared in a common component (or a shared superclass).
+3. Always use “private static final” constants instead of duplicating strings or other special values at different locations. Constants should be declared in the top member area of a class.
+
+
+### Design for Testability
+
+Code that is easily testable typically has good separation of concerns and is structured to be reusable outside the original context (by being easily reusable in tests).
+
+A good summary or problems / symptoms and recommended refactoring is in this PDF. 
```

Comment:
This sentence is kind of confusing, I thought "this pdf" meant the code quality document. Maybe "in the linked PDF"